src/oscar/apps/catalogue/product_attributes.py

```diff
@@ -19,14 +19,35 @@ def __setstate__(self, state):
         self.__dict__ = state
 
     def __init__(self, product):
-        self.product = product
-        self.refresh()
+        self._product = product
+        self._initialized = False
+
+    @property
+    def product(self):
+        return self._product
+
+    def initialize(self):
+        self._initialized = True
+        # initialize should not overwrite any values that have allready been set
+        attrs = self.__dict__
+        for v in self.get_values().select_related('attribute'):
+            attrs.setdefault(v.attribute.code, v.value)
 
     def refresh(self):
-        values = self.get_values().select_related('attribute')
-        for v in values:
+        for v in self.get_values().select_related('attribute'):
             setattr(self, v.attribute.code, v.value)
 
+    def __getattribute__(self, name):
```

Comment:
Wouldn't it be better to call the `initialize` method on the top here if it isn't yet initialized `(self._initialized)` instead of doing this in in the attribute error?

eg
```python
def __getattribute__(self, name):
    if not self._initialized:
        self.initialize()
    return super().__getattribute__(name)
```

Reply:
Most definately not. __getattribute__ is called on all attribute accesses, even methods and even on ``self._initalized``. The point is we only want to ``initialize`` only when we are trying to access a product attribute, and that is only when the ``AttributeError`` happens. In all other cases it is not a product attribute that is being accessed. We arre trying to delay the initialization as much as possible to avoid making useless queries.